File: src/engine/engine.js
// Public engine API: pick the best move for a position. JS port of kalah/engine.py.
//   mode: "solve" -> exact/perfect ; "play" -> time-limited search.
// A loaded book is consulted first in any mode (book moves are proven optimal).

import { MAX_DEPTH, TIME_BUDGET_MS } from './constants.js';
import { isTerminal, margin } from './board.js';
import { search, solve } from './search.js';
import { TranspositionTable } from './transposition.js';
import { bookMove } from './book.js';

export function bestMove(state, player, opts = {}) {
  if (isTerminal(state)) {
    return { move: -1, value: margin(state, player), source: 'terminal' };
  }
  const booked = bookMove(state, player);
  if (booked >= 0) return { move: booked, value: null, source: 'book' };

  if (opts.mode === 'solve') {
    const r = solve(state, player, { tt: opts.tt });
    r.source = 'solve';
    return r;
  }
  const r = search(state, player, {
    maxDepth: opts.maxDepth ?? MAX_DEPTH,
    timeMs: opts.timeMs ?? TIME_BUDGET_MS,
    tt: opts.tt,
  });
  r.source = 'search';
  return r;
}

export function analyze(state, player, opts = {}) {
  const tt = opts.tt || new TranspositionTable();
  const r = bestMove(state, player, { ...opts, tt });
  r.ttSize = tt.size;
  return r;
}

export { search, solve } from './search.js';
export {
  initialBoard,
  legalMoves,
  sow,
  isTerminal,
  cleanup,
  winner,
  margin,
  finalStores,
} from './board.js';
export {
  P1_TOP,
  P2_BOTTOM,
  OPPONENT,
  PITS,
  STORE,
  OPP_STORE,
  OPPOSITE,
  CELLS,
  WIN_THRESHOLD,
  TERMINAL_SCALE,
} from './constants.js';

File: kalah/engine.py
"""Public engine API: pick the best move for a position.

Thin orchestration over the search core, mirroring the reference engine's
``engine.js``. Two modes:
  * ``"solve"`` - exact, provably optimal (search to terminal).
  * ``"play"``  - consult the perfect-play book if loaded, else iterative
                  deepening to a time budget.
"""

from .constants import MAX_DEPTH, TIME_BUDGET_MS, P1_TOP, P2_BOTTOM
from .board import (
    initial_board,
    legal_moves,
    sow,
    is_terminal,
    cleanup,
    winner,
    margin,
)
from .search import search, solve
from .transposition import TranspositionTable
from . import book


def best_move(state, player, mode="play", **opts):
    """Return ``{"move", "value", "source", ...}`` for ``player`` in ``state``.

    ``mode="solve"`` always computes the exact answer. ``mode="play"`` prefers a
    book hit, then falls back to a time-limited search.
    """
    if is_terminal(state):
        return {"move": None, "value": margin(state, player), "source": "terminal"}

    if mode == "solve":
        result = solve(state, player, tt=opts.get("tt"))
        result["source"] = "solve"
        return result

    bm = book.book_move(state, player)
    if bm is not None:
        return {"move": bm, "value": None, "source": "book"}

    result = search(
        state,
        player,
        max_depth=opts.get("max_depth", MAX_DEPTH),
        time_budget_ms=opts.get("time_budget_ms", TIME_BUDGET_MS),
        tt=opts.get("tt"),
    )
    result["source"] = "search"
    return result


def analyze(state, player, mode="play", **opts):
    """Like :func:`best_move` but also reports transposition-table size."""
    tt = opts.pop("tt", None) or TranspositionTable()
    result = best_move(state, player, mode=mode, tt=tt, **opts)
    result["tt_size"] = tt.size
    return result


__all__ = [
    "best_move",
    "analyze",
    "initial_board",
    "legal_moves",
    "sow",
    "is_terminal",
    "cleanup",
    "winner",
    "margin",
    "search",
    "solve",
    "P1_TOP",
    "P2_BOTTOM",
]
